File: modules/data/thresh.lua
sptbl["thresh"] = {

    files = {
        module = "thresh.c",
        header = "thresh.h",
        example = "ex_thresh.c",
    },

    func = {
        create = "ut_thresh_create",
        destroy = "ut_thresh_destroy",
        init = "ut_thresh_init",
        compute = "ut_thresh_compute",
    },

    params = {
        optional = {
            {
                name = "thresh",
                type = "UTFLOAT",
                description = "Threshold level",
                default = 1.5
            },
            {
                name = "mode",
                type = "int",
                description ="Sets when to detect theshold crossings. 0 = from below. 1 = from above. 2 = from above/below",
                default = 0
            },
        }
    },

    modtype = "module",

    description = [[Trigger generator for signals that cross a given threshold. 
]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "input",
            description = "Input signal."
        },
    },

    outputs = {
        {
            name = "trig",
            description = "Produces a trigger when the input signal crosses the set threshold."
        },
    }

}

File: modules/thresh.c
#include <stdlib.h>
#include "utone.h"

int ut_thresh_create(ut_thresh **p)
{
    *p = malloc(sizeof(ut_thresh));
    return UT_OK;
}

int ut_thresh_destroy(ut_thresh **p)
{
    free(*p);
    return UT_OK;
}

int ut_thresh_init(ut_data *ut, ut_thresh *p)
{
    /* Initalize variables here. */
    p->init = 1;
    p->mode = 0;
    p->prev = 0;
    p->thresh = 0;
    return UT_OK;
}

int ut_thresh_compute(ut_data *ut, ut_thresh *p, UTFLOAT *in, UTFLOAT *out)
{
    if(p->init) {
        *out = 0;
        p->prev = *in;
        p->init = 0;
        return UT_OK;
    }

    switch((int)p->mode) {
        /* input signal goes above threshold */
        case 0:
            *out = (*in > p->thresh && p->prev <= p->thresh);
            break;

        /* input signal goes below threshold */
        case 1:
            *out = (*in < p->thresh && p->prev >= p->thresh);
            break;

        /* input signal goes below or above threshold */
        case 2:
            *out = (*in < p->thresh && p->prev >= p->thresh) ||
                (*in > p->thresh && p->prev <= p->thresh);
            break;

        default:
            return UT_NOT_OK;
    }

    p->prev = *in;
    
    return UT_OK;
}
